src/System.Security.Cryptography.Cng/tests/ImportExportTests.cs

```diff
@@ -46,5 +45,191 @@ public static void TestImportExportRoundTrip()
+                Assert.True(param.Curve is ECNamedCurve);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+                    param2 = newCng.ExportNamedCurveParameters(false);
+                    Assert.True(param2.Curve is ECNamedCurve);
+                    IsEqual(param, param2);
+                }
+            }
+        }
+
+        [Fact]
+        public static void TestExportImportPublicKeyPreWin10()
+        {
+            ECParameters param, param2;
+            ECDsaCng cng, newCng;
+
+            // ECDsaP256 should flow through without calling newer Win10 ecc functionality
+            using (cng = new ECDsaCng(CngKey.Create(CngAlgorithm.ECDsaP256)))
+            {
+                param = cng.ExportNamedCurveParameters(false);
+                Assert.True(param.Curve is ECNamedCurve);
+
+                // ECDsaP256 maps to nistP256
+                Assert.Equal("nistP256", ((ECNamedCurve)param.Curve).Oid.FriendlyName);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+                    param2 = newCng.ExportNamedCurveParameters(false);
+                    Assert.True(param2.Curve is ECNamedCurve);
+                    IsEqual(param, param2);
+                }
+            }
+        }
+
+        [Fact]
+        public static void TestExportImportPrimeCurve()
+        {
+            if (!EcDSACngTests.IsEccNamedCurveSupported())
+                return;
+
+            ECParameters param, param2;
+            ECDsaCng cng, newCng;
+
+            using (cng = new ECDsaCng(ECNamedCurve.nistP256))
+            {
+                param = cng.ExportExplicitCurveParameters(false);
+                Assert.True(param.Curve is ECPrimeCurve);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+                    param2 = newCng.ExportExplicitCurveParameters(false);
+                    IsEqual(param, param2);
+                }
+            }
+        }
+
+        [Fact]
+        public static void TestRegenKeyNegative()
+        {
+            if (!EcDSACngTests.IsEccNamedCurveSupported())
+                return;
+
+            ECParameters param;
+            ECDsaCng cng, newCng;
+
+            using (cng = new ECDsaCng(ECNamedCurve.brainpoolP160r1))
+            {
+                param = cng.ExportExplicitCurveParameters(true);
+                Assert.True(param.Curve is ECPrimeCurve);
+                Assert.NotEqual(param.D, null);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+
+                    // the curve name is not flowed on on explicit export\import (by design) so we are unable to generate new
+                    Assert.ThrowsAny<CryptographicException>(() => newCng.GenerateNewKey());
```

Comment:
No, I don't believe so. This code path calls NCryptCreatePersistedKey successfully (with algorithm="ECDSA") and then calls NCryptFinalizeKey but which returns invalid parameter because the curve name is not set (CngKey.Create.cs). A normal explicit import code path calls NCryptImportKey which does create the key fine given the explicit (full key) blob without curve name.

I am not sure why the curve name *must* be set, perhaps we should ask the Windows team.

A potential work-around is to automatically import the key instead, but that may be overkill.

Call stack:
>	system.security.cryptography.cng.dll!System.Security.Cryptography.CngKey.Create(System.Security.Cryptography.CngAlgorithm algorithm, string keyName, System.Security.Cryptography.CngKeyCreationParameters creationParameters) Line 57	C#
 	system.security.cryptography.cng.dll!Internal.Cryptography.CngAlgorithmCore.GetOrGenerateKey(int keySize, System.Security.Cryptography.CngAlgorithm algorithm, bool forceNewKey) Line 66	C#
 	system.security.cryptography.cng.dll!System.Security.Cryptography.ECDsaCng.GetKey(bool forceNewKey) Line 74	C#
 	system.security.cryptography.cng.dll!System.Security.Cryptography.ECDsaCng.GenerateNewKey() Line 46	C#